### PR TITLE
Scroll view and DataTable Widget

### PR DESCRIPTION
**Pull request by willmcgugan**
- Moved example to "old examples" directory. We can gradually start porting examples with new API.
- Copied LRUCache from Rich
- Implemented ScrollView widget
- Implemented DataTable widget

Data Table is still fairly rudimentary in terms of features. You add columns / rows and it will render it.

A future update will add a caret / selection and other features

https://user-images.githubusercontent.com/554369/174762713-acbc8491-773d-494f-906f-c87540e0f5bc.mov

### CODE REVIEW

File: src/textual/widgets/_datatable.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+@dataclass
+class Cell:
+    value: object
+
+
+class Header(Widget):
+    pass
+
+
+class DataTable(ScrollView, Generic[CellType]):
```
Comment (mrossinek) [inline src/textual/widgets/_datatable.py:39]: This class makes my heart beat faster :heart_eyes: 

I have a few questions/comments I wanted to leave her, I hope you don't mind :+1:

File: src/textual/widgets/_datatable.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    DataTable > .datatable--even-row {
+        background: $primary 10%;
+    }
+
+    DataTable > .datatable--highlight {
+        background: $secondary;
+        color: $text-secondary;
+    }
```
Comment (mrossinek) [inline src/textual/widgets/_datatable.py:68]: This `highlight` style looks very interesting. I don't see it being used right now but played around with it by storing a `current_row` on the `DataTable` instance which will use this instead of the even/odd style.
The `TableApp` can then update the `current_row` value when the user presses up/down (j/k, etc.). Was this potentially an intended use case here?
If so, it might make more sense to track the `current_row` on the App, rather than the internal table instance. This would require the option to change the style of a given row individually. Is something like that planned/in-scope?

File: src/textual/widgets/_datatable.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    def __init__(
+        self,
+        name: str | None = None,
+        id: str | None = None,
+        classes: str | None = None,
+    ) -> None:
+        super().__init__(name=name, id=id, classes=classes)
+        self.columns: list[Column] = []
+        self.data: dict[int, list[CellType]] = {}
```
Comment (mrossinek) [inline src/textual/widgets/_datatable.py:87]: Do you have any plans for the integration/support of `rich.table` objects? E.g. in a setting where some other code already uses rich and produces a `Table`, having the option of using that object directly and display it within a textual app would be amazing.
I am not sure whether such a tight integration is possible/planned/desired though :thinking:

File: src/textual/screen.py [hunk 4]
```diff
@@ -22,15 +22,16 @@
     from typing_extensions import Final
 
 # Screen updates will be batched so that they don't happen more often than 20 times per second:
-UPDATE_PERIOD: Final = 1 / 20
+UPDATE_PERIOD: Final = 1 / 60
```
Comment (mrossinek) [inline src/textual/screen.py:25]: I'm rewriting an application of mine, moving from an `ncurses` TUI + non-uniform CLI interface to rich+textual in order to attempt a unified UI for the user.
I couldn't help but notice that scrolling through the lines on a table in my old `ncurses` code is significantly smoother than scrolling in `sandbox/table.py`.

I am aware, that textual is very early stage but I wanted to ask whether you're aware of this, if you have plans on how this may be improved in the future, or in general how your experience in terms of performance has been thus far. Do you "frequently" compare to other TUI libraries?

**Review by mrossinek — COMMENTED** [review]
I'm very excited about this PR! I have been following its progress and am testing things out whenever I find the time. I wanted to drop by and leave some comments/questions for you, if you don't mind :slightly_smiling_face:

File: src/textual/widgets/_datatable.py (hunk 3 above)
Comment (willmcgugan) [reply src/textual/widgets/_datatable.py:87]: Any Rich renderable may be used in Textual, including Tables.

File: src/textual/screen.py (hunk 4 above)
Comment (willmcgugan) [reply src/textual/screen.py:25]: It's very smooth here, but I haven't started optimizing.

File: src/textual/widgets/_datatable.py (hunk 3 above)
Comment (mrossinek) [reply src/textual/widgets/_datatable.py:87]: I meant a closer relation between a `rich.table` and a `DataTable`. I.e. that textual would be able to treat the former identical to (or at least to some degree like) the latter.

File: src/textual/_compositor.py [hunk 5]
```diff
@@ -530,31 +546,34 @@ def _get_renders(
...
-
-        # Pretty sure we don't need to sort the bucket items
+        """Combine chops in to lines."""
+        from_iterable = chain.from_iterable
         segment_lines: list[list[Segment]] = [
-            sum(
-                [line for line in bucket.values() if line is not None],
-                [],
-            )
+            list(from_iterable(line for line in bucket.values() if line is not None))
```
Comment (willmcgugan) [inline src/textual/_compositor.py:571]: `chail.from_iterable` is faster than `sum`

File: src/textual/widgets/_datatable.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+        fixed, scrollable = self._render_row(row_index, line_no, base_style)
+        fixed_width = sum(column.width for column in self.columns[: self.fixed_columns])
+
+        fixed_line: list[Segment] = list(chain.from_iterable(fixed)) if fixed else []
+        scrollable_line: list[Segment] = list(chain.from_iterable(scrollable))
+
+        segments = fixed_line + line_crop(scrollable_line, x1 + fixed_width, x2, width)
+
+        # line = Segment.adjust_line_length(segments, width, style=base_style)
```
Comment (DrBenton) [inline src/textual/widgets/_datatable.py:278]: I reckon we can get rid of that commented code? :-)

File: src/textual/widgets/_datatable.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+        lines = [render_line(y, x1, x2, base_style) for y in range(y1, y2)]
+
+        for line_index, y in enumerate(range(y1, y2)):
+            if y - scroll_y < fixed_top_row_count:
+                lines[line_index] = fixed_lines[line_index]
+
+        return lines
+
+    def on_mouse_move(self, event):
+        print(self.get_style_at(event.x, event.y).meta)
```
Comment (DrBenton) [inline src/textual/widgets/_datatable.py:317]: this could probably be "decorated" with a `TODO` comment, as this is really not the kind of stuff we want to forget for the release? :sweat_smile:

File: tests/test_integration_layout.py [hunk 8]
```diff
@@ -19,7 +19,13 @@
 SCREEN_SIZE = Size(SCREEN_W, SCREEN_H)
 PLACEHOLDERS_DEFAULT_H = 3  # the default height for our Placeholder widgets
 
+# TODO: Brittle test
+# These tests are currently way to brittle due to the CSS layout not being final
+# They are also very hard to follow, if they break its not clear *what* went wrong
+# Going to leave them marked as "skip" for now.
 
+
+@pytest.mark.skip("brittle")
```
Comment (DrBenton) [inline tests/test_integration_layout.py:28]: The fact that these tests breaks probably means that there is a regression in the layout algorithm - which is their raison d'etre?
i.e. elements with a visible border contained within a 100 cells container with a scrollbar used to fill all the available space (`100 - 1 scrollbar - (border*2) = 97`), but they are now 1 cell smaller?
Screenshot from 2022-06-21 11-06-08

File: sandbox/table.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    def action_toggle_dark(self) -> None:
+        self.app.dark = not self.app.dark
+
+    def action_toggle_zebra(self) -> None:
+        self.table.zebra_stripes = not self.table.zebra_stripes
+
+    def action_exit(self) -> None:
+        from rich.style import Style
+
+        self.exit(Style._add_cache.cache_info())
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply sandbox/table.py:73]: Good spot. The result of a local copy of Rich.

File: src/textual/color.py [hunk 10]
```diff
@@ -235,7 +235,6 @@ def with_alpha(self, alpha: float) -> Color:
         r, g, b, _ = self
         return Color(r, g, b, alpha)
 
-    @lru_cache(maxsize=2048)
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/color.py:238]: The factor is a floating point from 0 to 1, derived from the current time. Basically it rarely got the same values twice, so did not benefit from caching.

File: src/textual/widgets/_datatable.py (hunk 7 above)
Comment (willmcgugan) [reply src/textual/widgets/_datatable.py]: I won't forget. Still working on this class...

File: src/textual/_node_list.py [hunk 11]
```diff
@@ -19,7 +19,8 @@ class NodeList:
 
     def __init__(self) -> None:
         # The nodes in the list
-        self._nodes: list[DOMNode] = []
+        self._nodes: list[Widget] = []
+        self._nodes_set: set[Widget] = set()
```
[the remark this answers is not in the document]
Comment (DrBenton) [reply src/textual/_node_list.py:23]: Ah, indeed! Forget me --> []  :sweat_smile:

**Review by darrenburns — CHANGES_REQUESTED** [review]

File: tests/test_integration_layout.py (hunk 8 above)
Comment (darrenburns) [reply tests/test_integration_layout.py:28]: Disagree from me on skipping this test and the reasoning behind it unless we foresee considerable changes to layout. The test done its job here. I don't think it's fair to call it brittle, and I don't agree that it's hard to follow or that it's hard to see what went wrong when it failed.

File: src/textual/_cache.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+    @overload
+    def get(self, key: CacheKey) -> Optional[CacheValue]:
+        ...
+
+    @overload
+    def get(
+        self, key: CacheKey, default: DefaultValue
+    ) -> Union[CacheValue, DefaultValue]:
+        ...
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/_cache.py]: It tells Mypy that if you call get with no default arg, the return type is `CacheValue` or None. And if you call it with a default arg, the return type is a union of `CacheValue` and `DefaultValue`

File: src/textual/_cache.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        Args:
+            key (CacheKey): Key
+            default (Optional[DefaultValue], optional): Default to return if key is not present. Defaults to None.
+
+        Returns:
+            Union[CacheValue, Optional[DefaultValue]]: Either the value or a default.
+        """
+        link = self.cache.get(key)
+        if link is None:
+            # Cache hit!
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/_cache.py]: Actually they were both wrong! Deleted those comments.

File: src/textual/_compositor.py [hunk 14]
```diff
@@ -316,7 +320,7 @@ def add_widget(
             container_size = container_region.size
 
             # Containers (widgets with layout) require adding children
-            if widget.is_container:
+            if widget.is_scrollable:
                 # The region that contains the content (container region minus scrollbars)
                 child_region = widget._arrange_container(container_region)
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/_compositor.py:325]: Gone for `_get_scrollable_region` since it returns a single Region.

File: src/textual/_cache.py [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+
+    __setitem__ = set
+
+    @overload
+    def get(self, key: CacheKey) -> CacheValue | None:
+        ...
+
+    @overload
+    def get(self, key: CacheKey, default: DefaultValue) -> CacheValue | DefaultValue:
+        ...
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/_cache.py:97]: t tells Mypy that if you call `get` with no default arg, the return type is CacheValue or None. And if you call it with a default arg, the return type is a union of CacheValue and DefaultValue

File: src/textual/_segment_tools.py [hunk 16]
```diff
@@ -7,14 +7,16 @@
...
 
-def line_crop(segments: list[Segment], start: int, end: int) -> list[Segment]:
+def line_crop(
+    segments: list[Segment], start: int, end: int, total: int
+) -> list[Segment]:
     """Crops a list of segments between two cell offsets.
 
     Args:
         segments (list[Segment]): A list of Segments for a line.
         start (int): Start offset
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/_segment_tools.py:17]: Will do

File: src/textual/_segment_tools.py [hunk 17]
```diff
@@ -35,6 +37,13 @@ def line_crop(segments: list[Segment], start: int, end: int) -> list[Segment]:
     else:
         return []
 
+    if end >= total:
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/_segment_tools.py:40]: Don't _think_ so. If total is 10, and end is 10, then we want this fast path for all the remaining Segments.

[the remark this answers is not in the document]
Comment (darrenburns) [reply src/textual/_segment_tools.py:40]: Think you're right.

File: src/textual/widgets/_datatable.py [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+@dataclass
+class Cell:
+    value: object
+
+
+class Header(Widget):
+    pass
+
+
+class DataTable(ScrollView, Generic[CellType]):
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/widgets/_datatable.py]: Done

File: src/textual/widgets/_data_table.py [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+        self._clear_caches()
+
+    def watch_fixed_rows(self, fixed_rows: int) -> None:
+        self._clear_caches()
+
+    def watch_zebra_stripes(self, zebra_stripes: int) -> None:
+        self._clear_caches()
+
+    def _update_dimensions(self) -> None:
+        max_width = sum(column.width for column in self.columns)
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/widgets/_data_table.py:155]: Should probably be called `total_width`.

[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/widgets/_data_table.py]: Updated and added docstring. Clear now?

File: tests/test_cache.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
+from __future__ import unicode_literals
+
+from textual._cache import LRUCache
+
+
+def test_lru_cache():
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply tests/test_cache.py:8]: Added a few more granular tests.

File: tests/test_integration_layout.py (hunk 8 above)
Comment (willmcgugan) [reply tests/test_integration_layout.py:28]: I do forsee considerable changes to layout and related APIs. I suspect there will be several iterations of update + test fix before it becomes stable. I find it hard to follow. 🤷‍♂️

**Comment by darrenburns** [comment]
> @darrenburns Happy with my reponses to your review?

Looks like there's still a few outstanding comments

**Comment by willmcgugan** [comment]
Github chose to hide them from me

File: src/textual/widget.py [hunk 21]
```diff
@@ -812,13 +830,22 @@ def layout(self) -> Layout:
...
         return self.styles.layout is not None or bool(self.children)
 
+    @property
+    def is_scrollable(self) -> bool:
+        """Check if this Widget may be scrolled.
+
+        Returns:
+            bool: True if this widget may be scrolled.
+        """
+        return self.is_container
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/widget.py:847]: No, they are distinct but related concepts. A container is a widget with children. When that is scrolled, the compositor adjusts the position of the children on screen.

A ScrollView is not a container, in that it has no children. But it can have scrollbars. A Scrollview updates its own region and doesn't rely on the compositor to scroll.

Make sense?

**Review by darrenburns — APPROVED** [review]

**Comment by willmcgugan** [comment]
🥳 Thanks @darrenburns @DrBenton
